packages/network/src/logic/InstructionSender.ts

```diff
@@ -0,0 +1,108 @@
+import _ from 'lodash'
+import { Logger } from '../helpers/Logger'
+import { Metrics } from '../helpers/MetricsContext'
+import { StreamIdAndPartition, StreamKey } from '../identifiers'
+import { TrackerServer } from '../protocol/TrackerServer'
+import { NodeId } from './Node'
+import { TopologyStabilizationOptions } from './Tracker'
+
+/**
+ * Instructions are collected to buffers and sent after a short delay. For each stream
+ * there is a separate buffer.
+ * 
+ * We use debouncing to delay the sending. It means that we send the buffered instructions 
+ * when either of these conditions is satisfied: 
+ * - the topology stabilizes: no new instructions has been formed for the stream in 
+ *   X milliseconds
+ * - the buffer times out: we have buffered an instruction for Y milliseconds
+ * 
+ * When an instruction is added to a stream buffer, it may overwrite an existing
+ * instruction in the buffer if the both instructions share the same nodeId. In that 
+ * situation we expect that the previous instruction is no longer valid (it has a lower
+ * counterValue) and can be ignored.
+ */
+
+const DEFAULT_TOPOLOGY_STABILIZATION: TopologyStabilizationOptions = {
+    debounceWait: 100,
+    maxWait: 2000
+}
+
+const logger = new Logger(module)
+
+export interface Instruction {
+    nodeId: NodeId,
+    streamKey: StreamKey,
+    newNeighbors: NodeId[],
+    counterValue: number
+}
+
+class StreamInstructionBuffer {
+    private instructions: Map<NodeId,Instruction> = new Map()
+    private debouncedOnReady: () => void
+
+    constructor(topologyStabilization: TopologyStabilizationOptions, onReady: () => void) {
+        this.debouncedOnReady = _.debounce(onReady, topologyStabilization.debounceWait, {
+            maxWait: topologyStabilization.maxWait
+        })
+    }
+
+    addInstruction(instruction: Instruction) {
+        // may overwrite an earlier instruction for the same node
+        this.instructions.set(instruction.nodeId, instruction)
+        this.debouncedOnReady()
+    }
+
+    getInstructions(): IterableIterator<Instruction> {
+        return this.instructions.values()
+    }
+}
+
+export class InstructionSender {
```

Comment:
Maybe InstructionDebouncer or Buffer would be a more accurate name

Reply:
Yes, if this class would do only the debouncing logic and not the actual sending. Now it encapsulates the whole send logic (not much) and specifically the fact that debouncing is used. The `Tracker` class is quite large, so maybe it is good we some details can be refactored out from that class.

One alternative would be to divide this class to `InstructionSender` + `InstructionDebouncer`. Maybe we should do that if/when the class grows.